improver/calibration/reliability_calibration.py

```diff
@@ -488,6 +488,289 @@ def process(self, cubes, coordinates=None):
         return result
 
 
+class ManipulateReliabilityTable(PostProcessingPlugin):
+    """
+    A plugin to manipulate the reliability tables to before they are used to
+    calibrate a forecast. x and y must be collapsed.
+    The result is a reliability diagram with monotonic observation frequency.
+
+    Steps taken are:
+
+    1. If non-monotonicity of the observation frequency is detected, try
+    combining a pair of bins that appear non-monotonic. Only a single pair of
+    bins are combined.
+
+    2. If non-monotonicity of the observation frequency remains after trying
+    to combine a single pair of bins, replace non-monotonic bins by assuming a
+    constant observation frequency.
+    """
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        return "<ManipulateReliabilityTable>"
+
+    @staticmethod
+    def _extract_reliability_table_components(reliability_table):
+        """Extract reliability table components from cube
+
+        Args:
+            reliability_table (iris.cube.Cube):
+                A reliability table to be manipulated.
+
+        Returns:
+            Tuple[numpy.ndarray, numpy.ndarray, numpy.ndarray, iris.coord.DimCoord]
+                Tuple containing the updated observation count,
+                forecast probability sum, forecast count and probability bin
+                coordinate.
+        """
+        observation_count = reliability_table.extract(
+            iris.Constraint(table_row_name="observation_count")
+        ).data
+        forecast_probability_sum = reliability_table.extract(
+            iris.Constraint(table_row_name="sum_of_forecast_probabilities")
+        ).data
+        forecast_count = reliability_table.extract(
+            iris.Constraint(table_row_name="forecast_count")
+        ).data
+        probability_bin_coord = reliability_table.coord("probability_bin")
+        return (
+            observation_count,
+            forecast_probability_sum,
+            forecast_count,
+            probability_bin_coord,
+        )
+
+    @staticmethod
+    def _sum_pairs(array, lower, upper):
+        """
+        Replace a pair of values in an array with their sum.
+
+        Args:
+            array (numpy.ndarray):
+                Array to be modified.
+            lower (int):
+                Lower index of pair.
+            upper (int):
+                Upper index of pair.
+
+        Returns:
+            numpy.ndarray:
+                Array where a pair of values has been replaced by their sum.
+        """
+        array[lower] = np.sum(array[lower : upper + 1])
+        return np.delete(array, upper)
```

Comment:
Just to mention, yes, we don't want non-contiguous pairs. I had made a modification to this in #1317, shown below, however, it might be preferable to only provide one index, as long as it's clear from the docstring which index of the pair is expected.

```
array[lower] = np.sum([array[lower], array[upper]])
```

Reply:
I think I've done what you both asked